### PR TITLE
Add Support for Storing Repo Labels

### LINKED ISSUE #1680: Support targeting repos with specific label like #hacktoberfest
Maintainer of CodeTriage here. I'm interested in adding a proof of concept for displaying repos with a given tag. Let's keep it simple and start with `#hacktoberfest` on one new page as the goal. We can do this as an experiment.

## Context

Every year Digital Ocean sends shirts to developers who get qualifying commits in open source during the month of October. 

In prior years this caused spam as projects not equipped to get new PRs would get an influx of work. To help the new rule is projects must have a`hacktoberfest` or `hacktoberfest-accepted` label.

From: https://hacktoberfest.com/participation/

```
Hacktoberfest is now opt-in for maintainers, so only contribute to projects that indicate they’re looking for Hacktoberfest PR/MRs.

Once your PR/MR has passed this check, we won’t check this again (unless your PR/MR fails a check before this, such as it being marked as spammy).
```

## Implementation

I'm thinking it could be fun for someone to write such a page for hacktoberfest. Basically wait until October 1st then send a PR. I'll be happy to `hacktoberfest-accepted` this feature.

It will need:

- Create a task to schedule importing the data: Recurring tasks are all in https://github.com/codetriage/CodeTriage/blob/05f248a9af1af718852fbcd14d746d878650324d/lib/tasks/schedule.rake. Adding a new task would probably be the easiest for now.
- Grab data about the github repo. For interactiong with github API I use https://github.com/zombocom/git_hub_bub. 
- This is the doc site for github API for repos https://docs.github.com/en/rest/repos/repos. 
- This is the doc site for github API for labels for a repo https://docs.github.com/en/rest/issues/labels#list-labels-for-a-repository.
- For now I think instead of trying to find and store ALL labels (which could get complicated if there's a lot of them) let's just work on storing those two `hacktoberfest` and `hacktoberfest-accepted`. The database schema for repos will need to be updated to include this extra data, or another table created.
- Make a page that makes a query to repos with those tags and displays them.

I think this is a big project it can be split into several PRs:

- Code to query the API and update the database
- Code to pull it the repos with the tags from the database and render them via a view. Add a codetriage.com/hacktoberfest page
- Code to filter and sort those hacktoberfest projects by language (similar to homepage). 

Could perhaps be split up other ways, but that's a rough start.

### PR DESCRIPTION
**Pull request by yez**
@schneems thanks for the opportunity to contribute! 

## Description
Adds two new tables: `repo_labels` and `labels` to store label data

Created a new service class to store and associate labels for repos
  if the same `label` exists on multiple `repos`, it does not make
  duplicate `label` records.

Add simple rake task to iterate through all `repos` `in_batches` and fetch one by one

## Related Issue

Partially resolved #1680 

## Motivation and Context

People should be able to target Hacktoberfest 

## How Has This Been Tested?

Some simple integration testing of the newly created PORO 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: app/services/repo_label_assigner.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    @repo = repo
+    url = ['repos', repo.user_name, repo.name, 'labels'].join('/')
+    @github_bub_response = GitHubBub.get(url)
+  end
+
+  def create_and_associate_labels!
+    return unless github_bub_response.success?
+
+    remote_labels.each do |label_hash|
+      label_name = label_hash['name'].downcase
```
Comment (yez) [inline app/services/repo_label_assigner.rb:24]: The issue stated that we wanted to only store `hacktoberfest` related labels. I'm happy to add a `next unless` line here if that's still the case but this is pretty simple as is.

File: lib/tasks/schedule.rake [hunk 2]
```diff
@@ -111,4 +111,11 @@ namespace :schedule do
     next unless Date.today.sunday?
     Rake::Task['sitemap:refresh'].invoke
   end
+
+  desc 'fetch and assign labels for repos'
+  task fetch_labels_and_assign: :environment do
+    Repo.find_each(batch_size: 100) do |repo|
+      RepoLabelAssigner.new(repo: repo).create_and_associate_labels!
```
Comment (yez) [inline lib/tasks/schedule.rake:118]: I considered creating a new background job here and enqueueing it for 10 repos at a time but worried about rate-limiting and not knowing the concurrency of background processing in production. I also wasn't sure what the total count of repos in production was.

File: test/unit/repo_labels_assigner_test.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class RepoLabelsAssignerTest < ActiveSupport::TestCase
```
Comment (yez) [inline test/unit/repo_labels_assigner_test.rb:5]: Outside of my comfort zone here (I'm used to RSpec) so if this needs to change please let me know!

File: lib/tasks/schedule.rake (hunk 2 above)
Comment (schneems) [reply lib/tasks/schedule.rake:118]: Repo count is on the main page it's ~6,000. The way we do API requests is that we rotate through everyone's tokens on the service for each request. https://github.com/codetriage/CodeTriage/blob/31bd0e7f109803b01d7dfcd553b59c08d8d9707c/config/initializers/git_hub_bub.rb#L48-L54 

It's a good idea not to make a bunch of extra requests, but we're not that constrained. If this is N API requests where N is the number of repos in the system, that's absolutely no problem.

**Review by schneems — APPROVED** [review]
This is awesome, thanks a ton!

**Comment by yez** [comment]
@schneems thanks for accepting! It seems like Hacktoberfest wants you to add `hacktoberfest-accepted` label to this PR for it to count unfortunately. I'll be sure to add the following PRs with the scoping functionality soon!

screenshot

**Comment by schneems** [comment]
Updated

**Pull request merged**